code/modules/power/singularity/emitter.dm

```diff
@@ -87,77 +104,185 @@
-//	add_load(1000)
-/*	if((severity == 1)&&prob(1)&&prob(1))
-		if(src.active)
-			src.active = 0
-			src.use_power = IDLE_POWER_USE	*/
-	return 1
+	var/toggle
+	if(active)
+		active = FALSE
+		toggle = "off"
+		shot_number = 0
+		fire_delay = maximum_fire_delay
+		investigate_log("turned <font color='red'>off</font> by [key_name(user)]", "singulo")
+	else
+		active = TRUE
+		toggle = "on"
+		investigate_log("turned <font color='green'>on</font> by [key_name(user)]", "singulo")
+
+	to_chat(user, "You turn [src] [toggle].")
+	message_admins("Emitter turned [toggle] by [key_name_admin(user)] in ([x], [y], [z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>)")
+	log_game("Emitter turned [toggle] by [key_name(user)] in [x], [y], [z]")
+	update_icon()
 
 /obj/machinery/power/emitter/attack_animal(mob/living/simple_animal/M)
 	if(ismegafauna(M) && anchored)
-		state = 0
+		state = EMITTER_NEEDS_WRENCH
 		anchored = FALSE
 		M.visible_message("<span class='warning'>[M] rips [src] free from its moorings!</span>")
 	else
 		..()
 	if(!anchored)
 		step(src, get_dir(M, src))
 
+/obj/machinery/power/emitter/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/card/id) || istype(I, /obj/item/pda))
+		if(emagged)
+			to_chat(user, "<span class='warning'>The lock seems to be broken.</span>")
+			return
+		if(allowed(user))
+			if(active)
+				locked = !locked
+				to_chat(user, "<span class='notice'>The controls are now [locked ? "locked" : "unlocked"].</span>")
+			else
+				locked = FALSE //just in case it somehow gets locked
+				to_chat(user, "<span class='warning'>The controls can only be locked when [src] is online!</span>")
+		else
+			to_chat(user, "<span class='warning'>Access denied.</span>")
+		return
+
+	if(exchange_parts(user, I))
+		return
+
+	return ..()
+
+/obj/machinery/power/emitter/wrench_act(mob/living/user, obj/item/I)
+	. = TRUE
+	if(active)
+		to_chat(user, "<span class='warning'>Turn off [src] first!</span>")
+		return
+	if(state == EMITTER_WELDED)
+		to_chat(user, "<span class='warning'>[src] needs to be unwelded from the floor!</span>")
+		return
+
+	if(state == EMITTER_NEEDS_WRENCH)
+		for(var/obj/machinery/power/emitter/E in loc)
+			if(E.anchored)
+				to_chat(user, "<span class='warning'>There is already an emitter here!</span>")
+				return
+		state = EMITTER_NEEDS_WELDER
+		anchored = TRUE
+		user.visible_message("[user] secures [src] to the floor.",
+			"You secure the external reinforcing bolts to the floor.",
+			"You hear a ratchet")
+	else
+		state = EMITTER_NEEDS_WRENCH
+		anchored = FALSE
+		user.visible_message("[user] unsecures [src] reinforcing bolts from the floor.",
+			"You undo the external reinforcing bolts.",
+			"You hear a ratchet")
+	playsound(src, I.usesound, I.tool_volume, TRUE)
+
+/obj/machinery/power/emitter/screwdriver_act(mob/living/user, obj/item/I) // Multitool menu NYI
+	. = TRUE
+	if(state != EMITTER_WELDED)
+		return
```

Comment:
Before you could screwdriver emitters open at any time, but now it has to be welded. Is this an intended change?

Reply:
I'm pretty sure there *was* a good reason for this, but if there was I've forgotten it now.
All seems to work fine with that removed anyway.